### PR TITLE
LayerMap::replace

### PR DESCRIPTION
**Pull request by koivunej**
Cc: #3486

Adds a method to replace a particular layer from the LayerMap for the purposes of remote layer download and layer eviction. In those use cases read lock on layer map needs to be released after initial search, but other operations could modify layermap before replacing thread gets to run.

I don't feel too strongly about the existing code refactorings. The actual usage is not intended to be fully resolved on this PR (needs more error handling unrelated to this replace functionality).

### CODE REVIEW

**Comment by hlinnaka** [comment]
Didn't read the patch, but +1 on having an explicit `LayerMap::replace` function. I think it will be more readable and less error-prone in the callers than the current way of calling insert+delete. Even if the replace function itself just does insert+delete.

This doesn't yet change the callers though, right?

**Comment by koivunej** [comment]
> This doesn't yet change the callers though, right?

Yes at least trying get a sense of this working first. Was thinking it could be done on a separate PR, since evict is not merged either. No need to combine the two.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:439]: Original bogus comment still.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:428]:
```suggestion
    /// Replaces a previous layer with a new layer value if the given closure returns true.
```

File: pageserver/src/tenant/layer_map.rs
Comment (problame) [inline pageserver/src/tenant/layer_map.rs:718]: the assertion below is gone

**Review by problame — CHANGES_REQUESTED** [review]
I have no experience with the new layermap, so, can't really comment on whether the changes are correct. Sure looks like you did your due diligence.

> EDIT: included a demo replacing the layer downloaded remove+insert, but the error handling and additional tests is probably a bit more work than I want to do on this PR.

Yeah, we're not regressing in any way by omitting the error handling.
Please add `warn!` messages though, so we can scrape the logs and see whether this case happens in practice.

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:718]: the assert should still be in place, but it's about l0 removal removing one, ot about this comparison, unless i misunderstand which assert?

Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:718]: aah the comment was left unupdated, fixing

**Comment by bojanserafimov** [comment]
I like the idea of a special replace method. Mostly because we can always replace without rebuilding anything.

One way to do that is to make the layer map return `LayerKey` instead of `Layer`, and have a separate `HashMap<LayerKey, Layer>` that we use to find the actual layer later. On `replace` we only need to update this hashmap. There might be a cleaner implementation.

But if rebuilds during replace are not a problem let me know and I'll review this as is.

**Comment by koivunej** [comment]
> One way to do that is to make the layer map return `LayerKey` instead of `Layer`, and have a separate `HashMap<LayerKey, Layer>` that we use to find the actual layer later. On `replace` we only need to update this hashmap. There might be a cleaner implementation.

I agree, this is a better idea for avoiding the rebuilds because through evictions there might be more downloads. The indirection will give the same API for pretty much free...

I will look into this tomorrow and ping you again. I think it might be possible to use the existing `BufferedHistoricLayerCoverage<Value>` as `Value = u64` and just have a `HashMap<u64, L>` on the `LayerMap`. Replace should just update this hashmap.

**Comment by bojanserafimov** [comment]
The indirection might be avoidable with some interior mutability sudoku, but I'm not sure if the complexity is worth it.

**Comment by koivunej** [comment]
> interior mutability

We could also have a flag for this on the layer side, meaning the replacement would happen completly outside of LayerMap, but I failed to find what were you exactly thinking of.

So at least two alternatives to get to constant time replacement:
1. use indirection at the layermap
2. use interior mutability within layers (also redesign how remote layers are represented)

> `HashMap<_, _>` indirection

I tried to peek at this, but I think the implementation is not trivial nor did I get any great ideas by now, perhaps with more focus time that'd be possible.

I think we should go ahead with this `replace` functionality as is on this PR, because while not optimal, it's much more correct than `remove_historic` and `insert_historic` without any feedback. @bojanserafimov

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: [d4742ed](https://github.com/neondatabase/neon/pull/3513/commits/d4742edae00dce4ff71b02784ba4440b0b3b28c0)

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:428]: Well, this probably should be rewritten altogether to include the "unless" cases.

Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs]: [1402120](https://github.com/neondatabase/neon/pull/3513/commits/14021202b7ae5588c8c14760d5da57fce7b0cc4f)

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:433]: These different conditions, to me, anyways, are another argument for having an enum as a result type instead of a bool.
Slap a `Debug` on that enum, and the call site in `timeline.rs` can print it in the `error!()`.
Better readability and much easier debugging if the replacement ever fails in prod.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:619]: This refactor causes distracting churn in this diff.
IIUC the only thing that actually changes was the addition of

```
        assert_eq!(version.delta_coverage.query(4), Some("Delta 1".to_string()));
```

below, correct?

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:689]: What's the point of trying evict & download with 3 non-overlapping layer keys in a row, instead of just one layer in the map?
Especially because the key `4` is hard-coded below?

IMO more interesting case would be a stack of three layers with identical key range but different LSNs.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:732]: suggestion: do the query that is done below after this as well, to assert we didn't modify the map state?

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:757]: would be nice to assert_eq the reason why the replace fails (=> enum return type)

File: pageserver/src/tenant/layer_map.rs
Comment (problame) [inline pageserver/src/tenant/layer_map.rs:322]: We don't have test coverage for this `if was_l0 { ... } else { ... }` branch, do we?

**Review by problame — COMMENTED** [review]
Here's a (relatively nitpicky review).

In addition to the remarks, this change makes it apparent that we don't remove the downloaded layer file from disk if the replacement fails. Also, we don't decrement the resident size gauge. The next access to the layer will trigger another on-demand download, which probably replaces the layer file, but increases the resident size gauge again.

That is a pre-existing bug, so, it's not necessarily your job to fix it.
It's been like this since Bojan's layer map rewrite.

Before that, we had the failing assert which would render the tenant (and pageserver unusable). Upon restart, timeline load would treat the downloaded layer file like any other layer file.
That behavior was better than what we have now.

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3280]: > If we fail to replace the layer, the next access will retry the on-demand download.

Ending up in the `Ok(false)` would mean that the layer we downloaded was no longer in the layermap. 

I think this retrying could only happen when there's another task already queued up by the semaphore waiting for the first attempt to fail to try downloading it again, but we don't diverge here, so the semaphore is closed and ... Something happens afterwards. I think the search should be restarted but I didn't find any way to signal that.

But this is exactly what I think should be handled on a follow-up (hence TODO).

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:318]: I don't agree: These are booleans, determined from the Layer::is_incremental. The downloaded layer name is already logged, twice if I remember correctly, and we keep a strong reference to it from the beginning of the download to the replacement attempt.

So this is just a sanity check that you don't try to replace an image layer with L0 delta layer.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:689]: > more interesting case would be a stack of three layers with identical key range but different LSNs.

They do have overlapping key ranges (overlap at 4 -- the hardcoded value), and different LSNs (see the `keys`). I think the test is exactly what you are hoping to see, please re-check.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:757]: It does not fail because we are in the state of having not-yet rebuilt insert to that key. Does this fail for you locally?

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:322]: I didn't look at coverage reports. Are you thinking it might overwrite a wrong index?

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:619]: Yes, it's in separate commit from the `map.get().expect("rebuilt")` changes.

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:314]: `was_l0` as in the to-be-replaced layer `expected`.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:433]:
```rust
enum Replacement<Value> {
  /// Old value was replaced with the given new value.
  Replaced {
    /// `true` if the key had a
    in_buffered: boolean
  },

  /// Key was not found in buffered or existing layers.
  NotFound,

  /// A removal for the replaced key has been buffered so no replacement was done.
  RemovalBuffered,

  /// Value was rejected by the given closure.
  Unexpected(Value),
}
```

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:732]: commit 19535141 (will push in a bit)

Also it doesn't really make sense to put something into the map with the same key, I'll remove rest of the test on a separate commit.

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3277]: How is `XXX` better than a normal TODO comment? I was hoping to leave figuring this out to an another issue.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:732]: Removing rest in a1934783

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:305]: Removed this duplicated comment in 38c861b40a7f, it's already documented under buffered.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:433]: [38c861b](https://github.com/neondatabase/neon/pull/3513/commits/38c861b40a7fcbce10b49e5b9190bc8036964b2e)

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: Included in [d4bbd9c](https://github.com/neondatabase/neon/pull/3513/commits/d4bbd9cfd750a6ad7d13d8501946cd8390a14042)

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: Included in [d4bbd9c](https://github.com/neondatabase/neon/pull/3513/commits/d4bbd9cfd750a6ad7d13d8501946cd8390a14042)

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:433]: Interesting borrowck failure, rust-analyzer doesn't show it (had began to rely on it).

Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:433]: Ok it's nll/polonius issue 1.

Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:433]: I'll just clone the arc.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs]: Resolving via the introduction of `enum Replacement`, still no assert_eq.

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: I'll see this CI run and put a todo!() there.

Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: Just making sure before I do the todo!() run -- there's an error for the case of l0 delta layer not being found from the layers, this is similar to the `assert_eq!(before, l0_deltas.len() - 1)`, but I just check it as a pre condition and not use an assert.

I think the `else` branch is not interesting, because nothing should be done for non-l0 delta layers.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (bojanserafimov) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:439]: not anymore

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (bojanserafimov) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:469]: Clones are only noticeable in querying code. For updates, go crazy.

File: pageserver/src/tenant/timeline.rs
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3278]: I guess it's just my background, it doesn't matter. Let's close this as hairsplitting.

File: pageserver/src/tenant/timeline.rs
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3281]: Yeah, not debating any of that. Still, it should be `error!()`, not `warn!()`. The resident size metric will be off in case there's a retry.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (bojanserafimov) [inline pageserver/src/tenant/layer_map/historic_layer_coverage.rs:629]: Thanks for the fix. Seems like there wasn't also an actual bug in the implementation, because I also queried it as an image :)

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:708]: :man_facepalming:  you're right.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (problame) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:638]: Ok, would be nice if you can keep them separate when merging, or merge that commit as an independent PR

File: pageserver/src/tenant/timeline.rs
Comment (bojanserafimov) [reply pageserver/src/tenant/timeline.rs:3278]: > I don't think we're consistent about this, but IMO TODO / todo!() is something that should be fixed before merging a PR.
XXX is something that is acceptable for now but needs to be revisited later.

I thought it was the opposite :) IMO what works best with github is self-reviewing and commenting over the todos you intend to fix now. Then github stops you from merging if there are unresolved conversations.

File: pageserver/src/tenant/layer_map.rs
Comment (problame) [reply pageserver/src/tenant/layer_map.rs]: I was using coverage in the broad sense here. All tests added by this PR are in `historic_layer_coverage.rs`.
So, the aspect of "need to remember to update `l0_delta_layers`" isn't covered by unit tests.

Which is probably fine, the other methods don't have such coverage either.

Feel free to mark this as resolved.

File: pageserver/src/tenant/timeline.rs
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3295]: Will the addresses be of any use to us when looking at the logs?
I think type info and layer file name would be more useful, as they could spark some initial investigation.

File: pageserver/src/tenant/layer_map.rs
Comment (bojanserafimov) [inline pageserver/src/tenant/layer_map.rs:148]:
```suggestion
    ///
    /// TODO replacement can be done without buffering and rebuilding layer map updates.
    ///      One way to do that is to add a layer of indirection for returned values, so
    ///      that we can replace values only by updating a hashmap.
    pub fn replace_historic(
```

Worth commenting, in case this becomes an issue and someone else trips on it.

**Review by problame — APPROVED** [review]
This LGTM, but let's get @bojanserafimov 's approval as well.

**Review by bojanserafimov — APPROVED** [review]
The layer map changes look good to me. I'll let the storage folks review `timeline.rs`

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: Aah yes, failed to think about unit tests because "pageserver code", but we can and should write some for the layermap. I'll add some for the add/replace/remove for non-l0 and l0.

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3295]: Good idea! Though I don't think there's any reason Layer's don't always impl Debug as well.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:629]: > Seems like there wasn't also an actual bug in the implementation, because I also queried it as an image :)

Exactly it only caused confusion :)

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:469]: Yeah, just API wise the `&Value` would had been nicer :)

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs]: c51adff8

File: pageserver/src/tenant/layer_map.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs]: Tests: [c4f4dfd](https://github.com/neondatabase/neon/pull/3513/commits/c4f4dfd82e582c89ebf722ae6bc3ff02be88375f) followed by some refactoring-with-tests :)

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3295]: Well, except for having a lot of fields I see.

File: pageserver/src/tenant/layer_map/historic_layer_coverage.rs
Comment (koivunej) [reply pageserver/src/tenant/layer_map/historic_layer_coverage.rs:639]: I don't think that it's that important, it's a minor test-misunderstanding after all. While I agree the follow up refactoring make the both a bit messy when looking at the first time.

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3295]: If it would have a different layer file name, it would also have a different `LayerKey` so it would not be found.

We could add `fn type_name() -> &'static str` to the trait, but I think that's what we should use `Debug` for... `short_id` is one way, but I don't think it could reveal anything we don't already have in the log, also it would be the same for Remote and Non-Remote.

Well, trying to figure something out.

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3281]: Changing this to `error!` as wished after rebase, also linking the created todo issue. I don't however think that this should be an error, at least per the quoted rfc:

- [x] not operating normally because we don't see how this could happen during normal operation
- [x] yes, unexpected
- [x] we continue anyway as if the replacement had worked (no change here)

I think errors were meant to be used only with error handling (likely on the TODO issue), which in this case will most likely be the get_reconstruct_data not finding anything, and postgres ending up looking at a zero page.

File: pageserver/src/tenant/timeline.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3295]: There is already a:

```rust
impl std::fmt::Debug for dyn Layer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Layer")
            .field("short_id", &self.short_id())
            .finish()
    }
}
```

Not sure I see what is the point of this... I'll get rid of it and include a full derived Debug, it'll repeat a lot of already known context but I cannot see a way around it.

Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:3295]: I'll do this Debug impls in a follow-up, I hate when every PR has to touch every file in the repo. Well, even 5 more.

**Pull request merged**
